test/e2e/cluster_upgrade.go

```diff
@@ -107,55 +214,93 @@ func testClusterUpgrade(framework Framework, svcName, podName string, upgrade fu
 		wg.Add(1)
 		defer wg.Done()
 
-		expectNoError(wait.Poll(poll, singleCallTimeout, func() (bool, error) {
+		if err := wait.Poll(poll, singleCallTimeout, func() (bool, error) {
 			r, err := httpClient.Get("http://" + ip)
 			if err != nil {
+				Logf("Error reaching %s: %v", ip, err)
 				return false, nil
 			}
 			if r.StatusCode < http.StatusOK || r.StatusCode >= http.StatusNotFound {
+				Logf("Bad response; status: %d, response: %v", r.StatusCode, r)
 				return false, nil
 			}
 			return true, nil
-		}))
+		}); err != nil {
+			// We log the error here because the test will fail at the very end
+			// because this validation runs in another goroutine. Without this,
+			// a failure is very confusing to track down because from the logs
+			// everything looks fine.
+			Errorf("Failed to contact service during master upgrade: %v", err)
```

Comment:
Kind of! I changed the expectNoError to a Failf because, assuming they work the same, Failf is better because we get to control the error message. (Not yet pushed.)

However, I think we still want a logging line here because that way we can see where the error occurred during the log. Without this, the Failf will just show an error at the very end of the test, after it looks like everything succeeded, and it's really confusing.